pkg/cloudprovider/providers/aws/instancecache.go

```diff
@@ -0,0 +1,308 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"sync"
+)
+
+type instanceCache struct {
+	cloud *Cloud
+
+	mutex  sync.RWMutex
+	latest *instanceCacheSnapshot
+}
+
+type awsInstanceID string
+
+func (i awsInstanceID) awsString() *string {
+	return aws.String(string(i))
+}
+
+type instanceCacheSnapshot struct {
+	timestamp    int64
+	byInstanceID map[awsInstanceID]*cachedInstance
+	byNodeName   map[types.NodeName]*cachedInstance
+}
+
+type cachedInstance struct {
+	ID awsInstanceID
+
+	// Values that cannot change are exported publicly
+	NodeName         types.NodeName
+	VpcID            string
+	AvailabilityZone string
+	InstanceType     string
+	SubnetID         string
+
+	cache *instanceCache
+
+	mutex sync.Mutex
+
+	state          *ec2.Instance
+	stateTimestamp int64
+}
+
+func (i *cachedInstance) DescribeInstance(cachePolicy *CachePolicy) (*ec2.Instance, error) {
+	snapshot := i.cache.latestSnapshot()
+
+	if snapshot != nil {
+		instance := snapshot.findCachedInstance(i.ID)
+		if instance != nil {
+			state := i.findLatestState(cachePolicy)
+			if state != nil {
+				return state, nil
+			} else {
+				glog.V(2).Infof("triggering instance refresh because instance state expired (instance %s, %s)", i.ID, cachePolicy.Name)
+			}
+		}
+	}
+
+	updated, err := i.cache.refreshAll(snapshot)
+	if err != nil {
+		return nil, err
+	}
+
+	instance := updated.findCachedInstance(i.ID)
+	if instance == nil {
+		return nil, cloudprovider.InstanceNotFound
+	}
+
+	state := i.findLatestState(nil)
+	if state != nil {
+		return state, nil
+	}
+
+	return nil, cloudprovider.InstanceNotFound
+}
+
+func (i *cachedInstance) findLatestState(cachePolicy *CachePolicy) *ec2.Instance {
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+
+	if i.state != nil {
+		if cachePolicy == nil || cachePolicy.IsValid(i.stateTimestamp) {
+			return i.state
+		} else {
+			glog.V(2).Infof("triggering refresh because cache policy %s expired", cachePolicy.Name)
+		}
+	}
+
+	return nil
+}
+
+func newInstanceCache(cloud *Cloud) *instanceCache {
+	return &instanceCache{
+		cloud: cloud,
+	}
+}
+
+func (c *instanceCache) latestSnapshot() *instanceCacheSnapshot {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	return c.latest
+}
+
+func (c *instanceCache) refreshAll(previous *instanceCacheSnapshot) (*instanceCacheSnapshot, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.latest != nil && c.latest != previous {
```

Comment:
The way I think about it is that two refreshAll requests come in at the same time, with the same `previous`.  One holds at the mutex lock while the other refreshes.  Then the second thread acquires the lock, but we have already updated.  So we want to skip the second refresh.